gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java

```diff
@@ -78,6 +84,12 @@
   private final ImmutableMap<String, String> metadata;
   private final boolean checksumsEnabled;
 
+  // Helper delegate for turning IOExceptions from API calls into higher-level semantics.
+  private ApiErrorExtractor errorExtractor = ApiErrorExtractor.INSTANCE;
+  // Determine if a given IOException is due to rate-limiting.
+  private RetryDeterminer<IOException> rateLimitedRetryDeterminer =
```

Comment:
I recalled that this will not work because `ApiErrorExtractor` relies on parsing of JSON API responses. Let's revert to your initial approach with retry method.